### PR TITLE
StartPage, cleanup/removal of collection items after entity deletion

### PR DESCRIPTION
**Pull request by janheise**
## Description

When an entity is deleted, preceding events in Recent Activity can no longer be displayed correctly and are removed now.

Also, deleted entities can no longer be a favorite or in the Last Opened list.

If a user is deleted, the Favorite & Last Opened collections for this user get removed, too.

/nocl

fixes https://github.com/Graylog2/graylog2-server/issues/14553 
fixes https://github.com/Graylog2/graylog2-server/issues/14403

## Motivation and Context

cleanup

## How Has This Been Tested?

manual testing, TODO: create tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by luk-kaminski — COMMENTED** [review]
What about some adding some tests for those new operations on MongoDB?

File: graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java [hunk 1]
```diff
@@ -105,4 +110,19 @@ public Optional<FavoritesForUserDTO> create(final FavoritesForUserDTO favorite,
             throw new IllegalStateException("Unable to create a Favorites collection, collection with this id already exists : " + favorite.id());
         }
     }
+
+    @Subscribe
+    public void removeFavoriteOnEntityDeletion(final RecentActivityEvent event) {
+        // if an entity is deleted, we can no longer see it in the lastOpened collection
+        if (event.activityType().equals(ActivityType.DELETE)) {
+            DBObject query = new BasicDBObject();
```
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java:118]: I'd also consider creating an index for field "items.id" if you agree that it is better to efficiently filter documents first, before executing the $pull operation.

File: graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java [hunk 2]
```diff
@@ -105,4 +110,19 @@ public Optional<FavoritesForUserDTO> create(final FavoritesForUserDTO favorite,
...
+            final DBObject modifications = new BasicDBObject("$pull", new BasicDBObject(FavoritesForUserDTO.FIELD_ITEMS, new BasicDBObject(FavoriteDTO.FIELD_ID, event.grn().entity())));
+            db.updateMulti(query, modifications);
+        }
+    }
+
+    @Subscribe
+    public void removeFavoriteEntityOnUserDeletion(final UserDeletedEvent event) {
+        db.remove(DBQuery.is(FavoritesForUserDTO.FIELD_USER_ID, event.userId()));
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java:126]: "user_id" field may need an index as well, what do you think?

File: graylog2-server/src/main/java/org/graylog/plugins/views/startpage/lastOpened/LastOpenedService.java [hunk 3]
```diff
@@ -16,34 +16,40 @@
...
 import org.mongojack.DBQuery;
 import org.mongojack.WriteResult;
 
 import javax.inject.Inject;
 import java.util.Optional;
 
-/*
- * TODO: remove entity, if a user is deleted?
- */
 public class LastOpenedService extends PaginatedDbService<LastOpenedForUserDTO> {
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog/plugins/views/startpage/lastOpened/LastOpenedService.java:40]: My comments from FavortiesService may apply to this service as well.

File: graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java (hunk 2 above)
Comment (janheise) [reply graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java:128]: :+1: done

File: graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (janheise) [reply graylog2-server/src/main/java/org/graylog/plugins/views/favorites/FavoritesService.java]: :+1: both done

File: graylog2-server/src/main/java/org/graylog/plugins/views/startpage/lastOpened/LastOpenedService.java (hunk 3 above)
Comment (janheise) [reply graylog2-server/src/main/java/org/graylog/plugins/views/startpage/lastOpened/LastOpenedService.java:40]: added same modifications on both classes

**Review by luk-kaminski — APPROVED** [review]
Looks good, great job!

**Pull request merged**